Subtask 2

St1ke opened the pull request.

St1ke committed:
[Subtask-2] Начало работы

St1ke committed:
[Subtask - 2]WIP Написал логику, требуется доработка исключений итд

St1ke committed:
[Subtask-2] WIP Исправление кода

1)Добавил исключения
2)Исправил несколько логических ошибок

St1ke committed:
[Subtask-2] Коммит для начала дрочева

1) Моральная устойчивость - отсутствует
2)Знания - отсутствуют

Tauders commented on Subtasks/README.md:
```diff
@@ -42,13 +42,25 @@
 4. Пользователь вводит, что сделать с числами (допустимые значения `+`, `-`, `/`, `*`)
 5. Вывод результата выполнения команды с прошлого шага на экран.
 
+2.1 Программа получает на вход строку с числами разделенными пробелами
+2.2 Программа 
+2.3 Программа распределяет эти числа в элементы массива
+numbers[0] = 4
+numbers[1] = 5
+numbers[2] = 6
+
+input[0] = 2
+input[1] = 1
+numbers[input[0] ]
+
+
```
Не меняй README.md, пожалуйста

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            }
+            Console.WriteLine("Числа выбраны, введите необходимое действие");
+            string actions = Console.ReadLine();
+            int total = choosenNumbers[0];
+            if (actions == "+")
+            {
+                for (i = 1; i <= counter; i++)
+                {
+                    total += choosenNumbers[i];
+                }
+            }
+            else if (actions == "-")
+            {
+                for(i = 1; i <= counter; i++)
+                {
+                    total -= choosenNumbers[i];
+                }
+            }
+            else if (actions == "/")
+            {
+                for (i = 1; i <= counter; i++)
+                {
+                    if (choosenNumbers[i] == 0)
+                    {
+                        Console.WriteLine("На ноль делить нельзя");
+                    }
+                    total /= choosenNumbers[i];
+                }
+            }
+            else if (actions == "*")
+            {
+                for (i = 1; i <= counter; i++)
+                {
+                    total *= choosenNumbers[i];
+                }
+            }
```
Вынеси в отдельный метод.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+        static void Main(string[] args)
+        {
+            Random rnd = new Random();
+            int value = rnd.Next(3, 7);
```
Не очень хорошее название - слишком абсрактное

St1ke replied:
![image](https://user-images.githubusercontent.com/83986957/150960291-eee0df29-f25f-4ab9-8e2d-a9d6bb24c749.png)
пофиксил

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            int[] convertedNumbers = new int[value];
+            Console.WriteLine($"Программа сгенерировала случайное число, {value}");
+            string[] numbers = new string[value];
+            string s = Console.ReadLine();
```
Не очень хорошее имя переменной - не говорящее.

St1ke replied:
про какую именно переменную речь? очень тяжело понять тк тут их целых три

Tauders replied:
`s`

St1ke replied:
![image](https://user-images.githubusercontent.com/83986957/150961206-b04b78ae-7cae-43fe-b1fb-52a211016e2e.png)
переименовал

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            int value = rnd.Next(3, 7);
+            int[] convertedNumbers = new int[value];
+            Console.WriteLine($"Программа сгенерировала случайное число, {value}");
+            string[] numbers = new string[value];
```
Зачем его заводить здесь? Перемести его объявление вниз.

St1ke replied:
Это не сложно, но зачем?  чтоб визуально понимать где вводится массив и используется?

Tauders replied:
В том числе, да. Потом, если бы код разростался, то между объявлением и фактическим использованием могло бы быть всё больше места, становилось бы всё непонятнее и т.д. 
Неплохая практика объявлять и использовать переменные как можно ближе к месту фактического использования. Из этого есть исключения, но пока можешь придерживаться такого подхода.

St1ke replied:
![image](https://user-images.githubusercontent.com/83986957/150962101-e208959d-05f4-43b7-bdd9-118cdd653342.png)
Исправлено

CodingFeles replied:
Ты создаёшь массив. ``` new string[value] ``` - выполняет реальную работу по выделению памяти. Ты потом нигде это не используешь и просто перезаписываешь содержимое массива. Вопрос: зачем тогда создавать его тут?
Если хочется, можешь оставить объявление, без инициализации. Но проще перенести это двумя строчками ниже.

Tauders replied:
Коммент был про `numbers` массив.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+                s = Console.ReadLine();
+                numbers = s.Split(new char[] { ' ' });
+            }
+            int i = 0;
```
Не очень хорошее название.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            int i = 0;
+            while (i < value)
+            {
+                bool firstSuccess = int.TryParse(numbers[i], out int newNumbers);
+                if (firstSuccess)
```
Сделай чуть иначе, разницы нет, но код будет меньше.

```cs
if (int.TryParse(numbers[i], out int newNumbers))
{
```

Или же поменяй имя переменной.

St1ke replied:
Не очень понимаю что "не так", звучит в духе, сделай что-нибудь, допустим имя переменной. "Первая проверка" вроде же норм?

Tauders replied:
"Первая проверка" не очень звучит, так как второй рядом нет? Первая для чего? И перевод то не такой: "Первый успех" звучит для названия переменной тоже странно - а где второй и т.д.

`isParsingSuccessful` или `isNumberParsed` или `isNumberParsedSuccessful` звучат лучше и понятнее.

Я же привёл способ, когда переменная отдельная вообще не нужна. Так ничего не поменяется, но код будет чуть меньше и не будет вопросов "если здесь есть переменная, то значит она для чего-то ещё нужна или какой в ней смысл".

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+                else
+                {
+                    Console.WriteLine("Некорректный ввод: Введите числовые значения");
+                    s = Console.ReadLine();
+                    numbers = s.Split(new char[] { ' ' });
```
Код дублируется 3 раза, вынеси в отдельный метод.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            int counter = 0;
+            string inputNumbers = "";
+            while (inputNumbers != "X")
+            {
+                inputNumbers = Console.ReadLine();
+                    if (inputNumbers != "X")
```
Форматирование.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            
+            int counter = 0;
+            string inputNumbers = "";
+            while (inputNumbers != "X")
```
Код дублируется, перепиши.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+                inputNumbers = Console.ReadLine();
+                    if (inputNumbers != "X")
+                    {
+                         bool secondSuccess = int.TryParse(inputNumbers, out int newInputNumbers);
+                         if (secondSuccess && newInputNumbers < value && newInputNumbers >= 0)
```
Или упрости, как выше, или поменяй название переменной.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+            }
+            if (counter == 0)
+            {
+                Console.WriteLine("Выход из программы");
```
Распиши, почему.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,104 @@
+                return;
+            }
+            Console.WriteLine("Числа выбраны, введите необходимое действие");
+            string actions = Console.ReadLine();
```
`actions` почему в множественном числе?

St1ke committed:
[Subtask-2]WIP Исправления различных ошибок

Исправлены все замечания, кроме создания отдельных методов.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,102 @@
+                return;
+            }
+            Console.WriteLine("Числа выбраны, введите необходимое действие");
+            string action = Console.ReadLine();
```
Что, если пользователь введёт не разрешённые действия?

St1ke committed:
[Subtask-2]WIP Исправление ошибок

1) Создал метод для выполнения действий с числами.
2)Добавил конструкцию switch/case

St1ke committed:
[Subtask-2] Корректировка кода

1)WIP Проверка ввода оператора действия в switch/case

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+            switch (action)
+            {
+                case "+":
+                    total = Summation(choosenNumbers, counter);
+                    break;
+                case "-":
+                    total = Subtraction(choosenNumbers, counter);
+                    break;
+                case "/":
+                    total = Division(choosenNumbers, counter);
+                    break;
+                case "*":
+                    total = Multiplication(choosenNumbers, counter);
```
Зачем во всех этих функциях `counter`? Потом он не используется в текущей функции и не понятно, зачем он там нужен.

St1ke replied:
ааа, не до конца понял смысл, ранее его значение используется в методе и принимается ,а тут чисто выходные и те что используются на выходе ?

St1ke replied:
![image](https://user-images.githubusercontent.com/83986957/151423738-0d5b9b07-8680-4903-b572-ad59611a3d4d.png)

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+                    Console.WriteLine("Вы ввели неверный оператор действия. Попробуйте ещё раз");
+                    break;
+            }
+            Console.WriteLine(total);
```
Это будет выводиться даже после того, как пользователь ввёл неверное действие?

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+                    total = Multiplication(choosenNumbers, counter);
+                    break;
+                default:
+                    Console.WriteLine("Вы ввели неверный оператор действия. Попробуйте ещё раз");
```
Добавь для тренировки возможно заново ввести действие.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+            }
+            return total;
+        }
+        public static string[] Stringsplit()
```
Не очень хорошее название функции - должно отражать действие и быть написать в `PascalCase`. 
Более удачное название будет в духе `ReadAndSplitUserInput`, или что-то.

Tauders commented on Subtasks/README.md:
```diff
@@ -48,7 +48,7 @@
 -> 4 5 6
 <- Числа введены, выберите доступные. X - выход
--> 0
+-> 2
```
Это что?

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+        public static string[] Stringsplit()
+        {
+            string userInput = Console.ReadLine();
+            string[] numbers = userInput.Split(new char[] { ' ' });
```
А почему это называется `numbers`? Здесь просто строки.

Tauders replied:
Можешь сделать здесь сразу возврат результатов выполнения функции `Split` и не придётся заводить отдельную переменную.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+            Random rnd = new Random();
+            int generatedNumber = rnd.Next(3, 7);
+            Console.WriteLine($"Программа сгенерировала случайное число, {generatedNumber}");
+            string[] numbers = Stringsplit();
```
Почему так называется переменная?

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+            
+            counter = 0;
+            string inputNumbers = "";
+            bool checkUserInput = false;
```
Название не очень. Переменная типа bool должна быть вопросом, ответ на который "да" или "нет".

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+        static void Main(string[] args)
+        {
+            Random rnd = new Random();
+            int generatedNumber = rnd.Next(3, 7);
```
Сделай имя более отражающим смысл.

Tauders commented on Subtasks/Evgeniy/Subtask-2/Subtask-2/Program.cs:
```diff
@@ -0,0 +1,137 @@
+                inputNumbers = Console.ReadLine();
+                if (inputNumbers != "X")
+                { 
+                    if (int.TryParse(inputNumbers, out int newInputNumbers) && newInputNumbers < generatedNumber && newInputNumbers >= 0)
```
Почему переменная `newInputNumbers` называется так? Там же одно число.

St1ke committed:
[Subtask-2] Final version

Исправил все замечания

St1ke committed:
[Subtask-2] Final version 2

St1ke committed:
[Subtask-2] Final version 3
